nucliadb/src/nucliadb/ingest/orm/knowledgebox.py

```diff
@@ -80,78 +85,140 @@ def new_unique_kbid() -> str:
-        # B/c with `Shards.similarity`, replaced by `model`
-        kb_shards.similarity = semantic_model.similarity_function
-
-        # if this KB uses a matryoshka model, we can choose a different
-        # dimension
-        if len(semantic_model.matryoshka_dimensions) > 0:
-            semantic_model.vector_dimension = choose_matryoshka_dimension(
-                semantic_model.matryoshka_dimensions  # type: ignore
-            )
-        kb_shards.model.CopyFrom(semantic_model)
+        try:
+            async with driver.transaction() as txn:
+                exists = await datamanagers.kb.get_kb_uuid(txn, slug=slug)
+                if exists:
+                    raise KnowledgeBoxConflict()
+                if kbid is None or kbid == "":
+                    kbid = str(uuid4())
 
-        kb_shards.release_channel = release_channel
+                if slug == "":
+                    slug = kbid
 
-        await datamanagers.cluster.update_kb_shards(txn, kbid=uuid, shards=kb_shards)
+                # Create in maindb
 
-        # shard creation will alter this value on maindb, make sure nobody
-        # uses this variable anymore
-        del kb_shards
-        shard_manager = get_shard_manager()
-        try:
-            await shard_manager.create_shard_by_kbid(txn, uuid)
-        except Exception as e:
-            await storage.delete_kb(uuid)
-            raise e
+                await datamanagers.kb.set_kbid_for_slug(txn, slug=slug, kbid=kbid)
 
-        return uuid
+                config = KnowledgeBoxConfig(
+                    title=title,
+                    description=description,
+                    slug=slug,
+                    migration_version=get_latest_version(),
+                )
+                await datamanagers.kb.set_config(txn, kbid=kbid, config=config)
+
+                # all KBs have the vectorset key initialized, although (for
+                # now), not every KB will store vectorsets there
+                await datamanagers.vectorsets.initialize(txn, kbid=kbid)
+
+                kb_shards = writer_pb2.Shards()
+                kb_shards.kbid = kbid
+                # B/c with Shards.actual
+                kb_shards.actual = -1
+                kb_shards.release_channel = release_channel
+
+                if semantic_model is not None:
+                    # bw/c we keep populating the deprecated fields for KBs with
+                    # only one vectorset
+
+                    # if this KB uses a matryoshka model, we can choose a different
+                    # dimension
+                    if len(semantic_model.matryoshka_dimensions) > 0:
+                        semantic_model.vector_dimension = choose_matryoshka_dimension(
+                            semantic_model.matryoshka_dimensions  # type: ignore
+                        )
+                    kb_shards.similarity = semantic_model.similarity_function
+                    kb_shards.model.CopyFrom(semantic_model)
```

Comment:
shouldn't we set the vectorsets here aswell, with the `__default__` vectorset_id?

Reply:
No, I'm not populating the /vectorsets key on KBs with only the default on purpose, so migration will be easier later. At this point, we don't know the model name (vectorset_id), so setting `__default__` here would require another check in the migration (KBs without the key, KBs with __default__ and KBs with vectorsets)